vendor/src/github.com/docker/libnetwork/portmapper/mapper.go

```diff
@@ -128,8 +128,10 @@ func (pm *PortMapper) MapRange(container net.Addr, hostIP net.IP, hostPortStart,
 	}
 
 	containerIP, containerPort := getIPAndPort(m.container)
-	if err := pm.forward(iptables.Append, m.proto, hostIP, allocatedHostPort, containerIP.String(), containerPort); err != nil {
-		return nil, err
+	if hostIP.To4() != nil {
+		if err := pm.forward(iptables.Append, m.proto, hostIP, allocatedHostPort, containerIP.String(), containerPort); err != nil {
+			return nil, err
+		}
```

Comment:
> When you submit to libnetwork, remember to also add the .To4() check around the pm.forward() in the following defer statement at line 138.

Done.

> In libnetwork your change to skip the iptables programming for v6 host IP, needed because we have no ip6tables package, may be ok, but would this work when daemon is started with --userland-proxy=false ? Can you give it a try ?

I’m not quite sure I get what you mean.

When starting docker with `--userland-proxy=false`, IPv6 connections don’t work at all:
```
$ docker run -p 80:80 nginx:1 &
$ curl -v -4 http://localhost:80/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.46.0
> Accept: */*
> 
< HTTP/1.1 200 OK
[…]
$ curl -v -6 http://localhost:80/
*   Trying ::1...
* Connected to localhost (::1) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.46.0
> Accept: */*
> 
[timeout]
```

I was under the impression that IPv6 requires the userland proxy.

In any case, my patch doesn’t change the outcome of this test.

If the test was not what you had in mind, might I ask you to check out this pull request and run the test yourself please?